helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java

```diff
@@ -77,15 +84,19 @@ public void process(ClusterEvent event) throws Exception {
       String instanceName = instance.getInstanceName();
       String instanceSessionId = instance.getEphemeralOwner();
 
-      // update pending messages
-      Map<String, Message> messages = cache.getMessages(instanceName);
-      Map<String, Message> relayMessages = cache.getRelayMessages(instanceName);
-      updatePendingMessages(instance, messages.values(), currentStateOutput, relayMessages.values(), resourceMap);
-
       // update current states.
       Map<String, CurrentState> currentStateMap = cache.getCurrentState(instanceName,
           instanceSessionId);
       updateCurrentStates(instance, currentStateMap.values(), currentStateOutput, resourceMap);
+
+      Map<String, Map<String, Message>> existingStaleMessages = cache.getStaleMessages();
+      currentStateOutput.setStaleMessageMap(existingStaleMessages);
+      // update pending messages
+      Map<String, Message> messages = cache.getMessages(instanceName);
+      Map<String, Message> relayMessages = cache.getRelayMessages(instanceName);
+      updatePendingMessages(instance, messages.values(), currentStateOutput,
+          relayMessages.values(), resourceMap, existingStaleMessages);
+      cache.setStaleMessages(currentStateOutput.getStaleMessageMap());
```

Comment:
The purpose of the cache is not for performance, but to keep all the historical stale messages. For example, assume a partition goes through M->S->M->S, and the first M->S message is not deleted even during the second M->S transition, if we only look at pending messages only, the first M->S message would be considered as non stale message, which causes confusion. With the cache, the stale messages would not be checked again.

Reply:
Then we need to define how the historical stale messages records are removed from the cache, or we will need to reboot the controller periodically, right?
Another concern is that if leadership switch happen, then you will still lose track of the first M->S message. So this is not a complete fix.

Reply:
The stale message is periodically cleaned up the same way as `pending message`, and the logic is in message generation phase. 
If controller switch happens, the in memory data is just lost right? Then all stale messages are automatically cleaned up. Any concern?

Reply:
I got the first point. Should be fine. Thanks for the explanation.

Regarding the leadership switch, the new leader won't know your cached value. So it will treat the previous message ( "first M->S message") as not stale. In general, if you have any logic depends on the cache, then the leadership switch will be troublesome.

Reply:
I see what you mean. I was talking about the case that message is gone from zk, but still in cache. So if the message hasn't been deleted from ZK yet, then yes, it'll stay there, and may cause confusion. I'm not sure whether this would cause a logic error. But I cannot think of other option unless we mark it in ZK.

Reply:
I think about it, I don't think it'll cause any problem even when controller switches. The message won't be processed if the target state is not the ideal state. The only problem is that the code won't automatically remove it for us.